pipeline doc - (#14414)
dotnet/docs: docs/standard/io/pipelines.md

<a name="solve"></a>

## What problem does System.IO.Pipelines solve
<!-- corner case doesn't MT (machine translate)   -->
Apps that parse streaming data are composed of boilerplate code having many specialized and unusual code flows. The boilerplate and special case code is complex and difficult to maintain.

`System.IO.Pipelines` was architected to:

* Have high performance parsing streaming data.
* Reduce code complexity.

The following code is typical for a TCP server that receives line-delimited messages (delimited by `'\n'`) from a client:

```csharp
async Task ProcessLinesAsync(NetworkStream stream)
{
    var buffer = new byte[1024];
    await stream.ReadAsync(buffer, 0, buffer.Length);
    
    // Process a single line from the buffer
    ProcessLine(buffer);
}
```

The preceding code has several problems:

* The entire message (end of line) might not be received in a single call to `ReadAsync`.
* It's ignoring the result of `stream.ReadAsync`. `stream.ReadAsync` returns how much data was read.

> * It doesn't handle the case where multiple lines are read in a single `ReadAsync` call.

Reviewer: I'd also consider mentioned that the method is always allocating a byte array each time we read data.

Author: > I'd also consider mentioning that the method is always allocating a byte array each time we read data.

Added

* It allocates a `byte` array with each read.

Typically, parsing takes more time than just copying blocks of data from the network:

* The reading thread gets ahead of the parsing thread.
* The reading thread has to either slow down or allocate more memory to store the data for the parsing thread.

For optimal performance, there's a balance between frequent pauses and allocating more memory.

To solve the preceding problem, the `Pipe` has two settings to control the flow of data:

* <xref:System.IO.Pipelines.PipeOptions.PauseWriterThreshold>: Determines how much data should be buffered before calls to <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> pause.
* <xref:System.IO.Pipelines.PipeOptions.ResumeWriterThreshold>: Determines how much data the reader has to observe before calls to `PipeWriter.FlushAsync` resume.

![Diagram with ResumeWriterThreshold and PauseWriterThreshold](./media/pipelines/resume-pause.png)

<xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A?displayProperty=nameWithType>:

* Returns an incomplete `ValueTask<FlushResult>` when the amount of data in the `Pipe` crosses `PauseWriterThreshold`.
* Completes `ValueTask<FlushResult>` when it becomes lower than `ResumeWriterThreshold`.

> Two values are used to prevent rapid cycling, which can occur if one value is used.

Reviewer: I'd give an example of the rapid cycling if the same value is used for the threshold.

Author: @jkotalik  please suggest verbiage.

### Reading multiple messages

The following code reads all messages from a `PipeReader` and calls `ProcessMessageAsync` on each.

[!code-csharp[MyConnection1](~/samples/snippets/csharp/pipelines/MyConnection1.cs?name=snippet)]

### Cancellation

`PipeReader.ReadAsync`:

* Supports passing a <xref:System.Threading.CancellationToken>.
* Throws an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.
* Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` causes the current or next call to `PipeReader.ReadAsync` to return a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.

[!code-csharp[MyConnection](~/samples/snippets/csharp/pipelines/MyConnection.cs?name=snippet)]

<a name="gotchas"></a>

### PipeReader common problems

> * Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed to `examined` but not `consumed`.

Reviewer 2: ```suggestion
* Passing the wrong values to `consumed` or `examined` may result in reading already read data.
```

<a name="pwcp"></a>

### PipeWriter common problems

* <xref:System.IO.Pipelines.PipeWriter.GetSpan%2A> and <xref:System.IO.Pipelines.PipeWriter.GetMemory%2A> return a buffer with at least the requested amount of memory. **Don't** assume exact buffer sizes.
* There's no guarantee that successive calls will return the same buffer or the same-sized buffer.
* A new buffer must be requested after calling <xref:System.IO.Pipelines.PipeWriter.Advance%2A> to continue writing more data. The previously acquired buffer can't be written to.
* Calling `GetMemory` or `GetSpan` while there's an incomplete call to `FlushAsync` isn't safe.
* Calling `Complete` or `CompleteAsync` while there's unflushed data can result in memory corruption.

## IDuplexPipe

The <xref:System.IO.Pipelines.IDuplexPipe> is a contract for types that support both reading and writing. For example, a network connection would be represented by an `IDuplexPipe`.

 Unlike `Pipe` which contains a `PipeReader` and a `PipeWriter`, `IDuplexPipe` represents a single side of a full duplex connection. That means what is written to the `PipeWriter` will not be read from the `PipeReader`.

## Streams

> When reading or writing stream data, you typically read data using a de-serializer and write data using a serializer. Most of these read and write stream APIs have a `Stream` parameter. To make it easier to integrate with these existing APIs, `PipeReader` and `PipeWriter` expose an <xref:System.IO.Pipelines.PipeReader.AsStream%2A>.  <xref:System.IO.Pipelines.PipeWriter.AsStream%2A> returns a `Stream` implementation around the `PipeReader` or `PipeWriter`.

Reviewer: Also mention that there are PipeReader.Create and PipeWriter.Create methods to convert a stream into a PipeReader/PipeWriter.

Reviewer 3: This is on me. I need to expand on the Stream interop with examples. I'll do this in a different PR

---
title: "I/O pipelines - .NET"

> ms.date: "10/01/2019"

Reviewer 4: one thing that is missing is the description metadata

---
title: "I/O pipelines - .NET"

> description: Shows how to avoid problems and efficiently use I/O pipelines in .NET

Reviewer 4: Probably focusing on usage first would be better?

```suggestion
description: Learn how to efficiently use I/O pipelines in .NET and avoid problems in your code.
```

author: rick-anderson
ms.author: riande
---
# System.IO.Pipelines in .NET

<xref:System.IO.Pipelines> is a new library that is designed to make it easier to do high-performance I/O in .NET. It’s a library targeting .NET Standard that works on all .NET implementations.

<a name="solve"></a>

## What problem does System.IO.Pipelines solve
<!-- corner case doesn't MT (machine translate)   -->
Apps that parse streaming data are composed of boilerplate code having many specialized and unusual code flows. The boilerplate and special case code is complex and difficult to maintain.

`System.IO.Pipelines` was architected to:

* Have high performance parsing streaming data.
* Reduce code complexity.

The following code is typical for a TCP server that receives line-delimited messages (delimited by `'\n'`) from a client:

```csharp
async Task ProcessLinesAsync(NetworkStream stream)
{
    var buffer = new byte[1024];
    await stream.ReadAsync(buffer, 0, buffer.Length);
    
    // Process a single line from the buffer
    ProcessLine(buffer);
}
```

The preceding code has several problems:

* The entire message (end of line) might not be received in a single call to `ReadAsync`.

> * It's ignoring the result of `stream.ReadAsync`. `stream.ReadAsync` returns how much data was read.

Reviewer 2: It's *not* at all typical to ignore the result of `stream.ReadAsync`.

Reviewer 2: Nvm. It's not typical, but it's a fine straw man.

* Buffer the incoming data until a new line is found.
* Parse all the lines returned in the buffer.
* It's possible that the line is bigger than 1 KB (1024 bytes). The code needs to resize the input buffer a complete line is found.

  * If the buffer is resized, more buffer copies are made as longer lines appear in the input.
  * To reduce wasted space, compact the buffer used for reading lines.

* Consider using buffer pooling to avoid allocating memory repeatedly.
* The following code address some of these problems:

[!code-csharp[](~/samples/snippets/csharp/pipelines/ProcessLinesAsync.cs?name=snippet)]

The previous code is complex and doesn't address all the problems identified. High-performance networking usually means writing very complex code to maximize performance. `System.IO.Pipelines` was designed to make writing this type of code easier.

## Pipe

The <xref:System.IO.Pipelines.Pipe> class can be used to create a `PipeWriter/PipeReader` pair. All data written into the `PipeWriter` is available in the `PipeReader`:

[!code-csharp[](~/samples/snippets/csharp/pipelines/Pipe.cs?name=snippet2)]

<a name="pbu"></a>

### Pipe basic usage

> [!code-csharp[](~/samples/snippets/csharp/pipelines/Pipe.cs?name=snippet)]

Reviewer 2: Shouldn't FillPipeAsync complete the PipeWriter with an exception if Socket.ReceiveAsync() throws? Completing the writer gracefully seems like bad practice.

There are two loops:

* `FillPipeAsync` reads from the `Socket` and writes to the `PipeWriter`.

> * `ReadPipeAsync` reads from the `PipeReader` and parses incoming lines.

Reviewer 2: I noticed mixed usage of `Complete` and `CompleteAsync` in this doc. These two methods call `Complete`, but `CompleteAsync` is also mentioned in a few places.

Shouldn't we use `CompleteAsync` everywhere to encourage people to use it and avoid possible sync-over-async bugs?

Reviewer: Agreed. We should deprecate Complete IMO.

Author: ```
async Task FillPipeAsync(Socket socket, PipeWriter writer)
{
// code omitted
// Tell the PipeReader that there's no more data coming.
    writer.Complete();   // <--- deprecate Complete IMO.
}

async Task ReadPipeAsync(PipeReader reader)
{
// code omitted

    // Mark the PipeReader as complete.
    reader.Complete();                              // <-   deprecate Complete IMO.
}
```

I'm not allowed to modify code. If one of your are brave, can you update that at https://github.com/dotnet/samples/tree/master/snippets/csharp/pipelines ?

Reviewer 2: Done. https://github.com/dotnet/samples/pull/1624

Reviewer 2: Done. https://github.com/dotnet/samples/pull/1624

  * A boolean `IsCompleted` that indicates if the end of data (EOF) has been reached. 

After finding the end of line (EOL) delimiter and parsing the line:

* The logic processes the buffer to skip what's already processed.
* `PipeReader.AdvanceTo` is called to tell the `PipeReader` how much data has been consumed and examined.

The reader and writer loops end by calling `Complete`. `Complete` lets the underlying Pipe release the memory it allocated.

### Backpressure and flow control

Ideally, reading and parsing work together:

* The writing thread consumes data from the network and puts it in buffers.
* The parsing thread is responsible for constructing the appropriate data structures.

Typically, parsing takes more time than just copying blocks of data from the network:

* The reading thread gets ahead of the parsing thread.
* The reading thread has to either slow down or allocate more memory to store the data for the parsing thread.

For optimal performance, there's a balance between frequent pauses and allocating more memory.

To solve the preceding problem, the `Pipe` has two settings to control the flow of data:

> * <xref:System.IO.Pipelines.PipeOptions.PauseWriterThreshold>: Determines how much data should be buffered before calls to <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> pauses.

Reviewer 2: Nit:

```suggestion
* <xref:System.IO.Pipelines.PipeOptions.PauseWriterThreshold>: Determines how much data should be buffered before calls to <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> pause.
```

Reviewer 2: It might also be worth calling out the unit is bytes.

> * <xref:System.IO.Pipelines.PipeOptions.ResumeWriterThreshold>: Determines how much data should be buffered before calls to `PipeWriter.FlushAsync` pauses.

Reviewer 2: ```suggestion
* <xref:System.IO.Pipelines.PipeOptions.ResumeWriterThreshold>: Determines how much data should be buffered before calls to `PipeWriter.FlushAsync` resumes.
```

> The `ResumeWriterThreshold` controls how much the reader has to observe before writing can resume.

Reviewer 2: This seem redundant given the above line.

* Completes `ValueTask<FlushResult>` when it becomes lower than `ResumeWriterThreshold`.

Two values are used to prevent rapid cycling, which can occur if one value is used.

### Examples

```csharp
// The Pipe will start returning incomplete tasks from FlushAsync until
// the reader examines at least 5 bytes.
var options = new PipeOptions(pauseWriterThreshold: 10, resumeWriterThreshold: 5);
var pipe = new Pipe(options);
```

### PipeScheduler

Typically when using `async` and `await`, asynchronous code resumes on either on a <xref:System.Threading.Tasks.TaskScheduler> or on the current  <xref:System.Threading.SynchronizationContext>.

When doing I/O, it's important to have fine-grained control over where the I/O is performed. This control allows taking advantage of CPU caches effectively. Efficient caching is critical for high-performance apps like web servers. <xref:System.IO.Pipelines.PipeScheduler> provides control over where asynchronous callbacks run. By default:

* The current <xref:System.Threading.SynchronizationContext> is used.
* If there's no `SynchronizationContext`, it uses the thread pool to run callbacks.

> [!code-csharp[](~/samples/snippets/csharp/pipelines/Program.cs?name=snippet)]

Reviewer 2: Should we say anything about the existence of `PipeScheduler.ThreadPool` and `PipeScheduler.Inline`? And the fact that `PipeScheduler.ThreadPool` is the default?

Author: @halter73  can you suggest text? Something like

`[PipeScheduler.ThreadPool](xref:System.IO.Pipelines.PipeScheduler.ThreadPool)`  is the `<xref:System.IO.Pipelines.PipeScheduler>` implementation that queues callbacks to thread pool. `PipeScheduler.ThreadPool` is the default and generally the best choice.  `[PipeScheduler.Inline](xref:System.IO.Pipelines.PipeScheduler.Inline)`  can cause unintended consequences such as deadlocks.

Reviewer 2: That looks great.

### Pipe reset

It's frequently efficient to reuse the `Pipe` object. To reset the pipe, call <xref:System.IO.Pipelines.PipeReader> <xref:System.IO.Pipelines.Pipe.Reset%2A> when both the `PipeReader` and `PipeWriter` are complete.

## PipeReader

<xref:System.IO.Pipelines.PipeReader> manages memory on the caller's behalf. **Always** call <xref:System.IO.Pipelines.PipeReader.AdvanceTo%2A?displayProperty=nameWithType> after calling <xref:System.IO.Pipelines.PipeReader.ReadAsync%2A?displayProperty=nameWithType>. This lets the `PipeReader` know when the caller is done with the memory so that it can be tracked. The `ReadOnlySequence<byte>` returned from `PipeReader.ReadAsync` is only valid until the call the `PipeReader.AdvanceTo`:

* It's illegal to use `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo`.

> * Using `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` is undefined.

Reviewer 2: Aren't both of the above bullet points saying basically the same thing?

`PipeReader.AdvanceTo` takes two <xref:System.SequencePosition> arguments:

* The first argument determines how much memory was consumed.
* The second argument determines how much of the buffer was observed.

> Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the unobserved data.

Reviewer 2: ```suggestion
Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the observed *and* unobserved data, but not data that has already been consumed.
```

> Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the unobserved data.

Reviewer 2: ```suggestion
Marking data as consumed means that the pipe can return the memory to the underlying buffer pool. Marking data as observed controls what the next call to `PipeReader.ReadAsync` does. Marking everything as observed means that the next call to `PipeReader.ReadAsync` won't return until there's more data written to the pipe. Any other value will make the next call to `PipeReader.ReadAsync` return immediately with the observed *and* unobserved data, but data that has already been consumed.
```

I know the point of the last sentence is to communicate that the next call to ReadAsync will always return immediately of you didn't observe everything in the last call, but se should make it very clear that ReadResult includes data that has already been observed. To this point in the documentation, I don't think it's that clear.

Reviewer 3: Feel free to use the suggested edit feature 😬

Reviewer 2: I thought I did. Didn't I?

Author: > I thought I did. Didn't I?

Yes, I took it in a batch.

[!code-csharp[ReadSingleMsg](~/samples/snippets/csharp/pipelines/ReadSingleMsg.cs?name=snippet)]

The preceding code:

* Parses a single message.
* Updates the consumed `SequencePosition` and examined `SequencePosition` to point to the start of the trimmed input buffer.

The two `SequencePosition` arguments are updated because `TryParseMessage` removes the parsed message from the input buffer. Generally, when parsing a single message from the buffer, the examined position should be one of the following:

* The end of the message.
* The end of the received buffer if no message was found.

The single message case has the most potential for errors. Passing the wrong values to *examined* can result in an out of memory exception or an infinite loop. For more information, see the [PipeReader common problems](#gotchas) section in this article.

### Reading multiple messages

The following code reads all messages from a `PipeReader` and calls `ProcessMessageAsync` on each.

[!code-csharp[MyConnection1](~/samples/snippets/csharp/pipelines/MyConnection1.cs?name=snippet)]

### Cancellation

`PipeReader.ReadAsync`:

* Supports passing a <xref:System.Threading.CancellationToken>.

> * Throws in an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.

Reviewer 2: Nit:

```suggestion
* Throws an <xref:System.OperationCanceledException> if the `CancellationToken` is canceled while there's a read pending.
```

> * Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` returns a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.

Reviewer 2: ```suggestion
* Supports a way to cancel the current read operation via <xref:System.IO.Pipelines.PipeReader.CancelPendingRead%2A?displayProperty=nameWithType>, which avoids raising an exception. Calling `PipeReader.CancelPendingRead` causes the current or next call to `PipeReader.ReadAsync` to return a <xref:System.IO.Pipelines.ReadResult> with `IsCanceled` set to `true`. This can be useful for halting the existing read loop in a non-destructive and non-exceptional way.
```

[!code-csharp[MyConnection](~/samples/snippets/csharp/pipelines/MyConnection.cs?name=snippet)]

<a name="gotchas"></a>

### PipeReader common problems

> * Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed.

Reviewer 2: ```suggestion
* Passing the wrong values to `consumed` or `examined` may result in reading already read data. For example, passing a `SequencePosition` that was already processed to `examined` but not `consumed`.
```

I'm not sure quite what was intended by the second sentence. It might be better to remove the second sentence entirely.

Reviewer 3: Yea it may make sense to remove it as this was written before I had the bad samples. Those show the cases more clearly

* Passing `buffer.End` as examined may result in:

  * Stalled data
  * Possibly an eventual Out of Memory (OOM) exception if data isn't consumed. For example, `PipeReader.AdvanceTo(position, buffer.End)` when processing a single message at a time from the buffer.

* Passing the wrong values to `consumed` or `examined` may result in an infinite loop. For example, `PipeReader.AdvanceTo(buffer.Start)` if `buffer.Start` hasn't changed will cause the next call to `PipeReader.ReadAsync` to return immediately before new data arrives.

> * Passing the wrong values to `consumed` or `examined` may result in infinite buffering (eventual OOM). For example, `PipeReader.AdvanceTo(buffer.Start, buffer.End)` unconditionally when processing a single message at a time from the buffer.

Reviewer 2: This seem redundant. Unconditionally passing `buffer.End` to examined leading to OOM was already mentioned 2 lines above.

* Using the `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` may result in memory corruption (use after free).
* Failing to call `PipeReader.Complete/CompleteAsync` may result in a memory leak.

> * Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this in the wrong order could result in an infinite loop.

Reviewer 2: Data loss I get. But how does checking IsEmpty and IsCompleted in the wrong order lead to an infinite loop?

Reviewer 2: I see after looking at the "Infinite loop" problematic code. It's not checking them in the wrong order that's problematic, it's expecting *both* to be true before you exit that can lead to an infinite loop given a stream ends with a partial message.

Reviewer 2: ```suggestion
* Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this incorrectly could result in an infinite loop.
```

Let's just go with that.

Reviewer 2: > Which sentence is that?

That quote was from [another comment](https://github.com/dotnet/docs/pull/14414#discussion_r333307312), but
I was referring to "For example, passing a `SequencePosition` that was already processed."

You applied the edit I suggested which improved it slightly, but I think that even the new expanded version is unclear and unnecessary. I'll suggest another edit to just remove it.

[!code-csharp[MyPipeWriter](~/samples/snippets/csharp/pipelines/MyPipeWriter.cs?name=snippet)]

The previous code:

* Requests a buffer of at least 5 bytes from the `PipeWriter` using <xref:System.IO.Pipelines.PipeWriter.GetSpan%2A>.
* Writes bytes for the ASCII string `"Hello"` to the returned `Span<byte>`.
* Calls <xref:System.IO.Pipelines.PipeWriter.Advance%2A> to indicate how many bytes were written to the buffer.
* Flushes the `PipeWriter`, which sends the bytes to the underlying device.

The previous method of writing uses the buffers provided by the `PipeWriter`. Alternatively, <xref:System.IO.Pipelines.PipeWriter.WriteAsync%2A?displayProperty=nameWithType>:

* Copies the existing buffer to the `PipeWriter`.
* Calls `GetSpan`, `Advance` as appropriate and calls <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A>.

[!code-csharp[MyPipeWriter#2](~/samples/snippets/csharp/pipelines/MyPipeWriter.cs?name=snippet2)]

### Cancellation

> <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> supports passing a <xref:System.Threading.CancellationToken>. Passing a `CancellationToken` results in an `OperationCanceledException` if the token is canceled while there's a flush pending. `PipeWriter.FlushAsync` supports a way to cancel the current flush operation via <xref:System.IO.Pipelines.PipeWriter.CancelPendingFlush%2A?displayProperty=nameWithType> without raising an exception. Calling `PipeWriter.CancelPendingFlush` returns a <xref:System.IO.Pipelines.FlushResult> with `IsCanceled` set to `true`. This can be useful for halting the yielding flush in a non-destructive and non-exceptional way.

Reviewer 2: This should be changed similarly to the PipeReader cancellation section to indicate that it's Flush/WriteAsync that returns a canceled flush result instead of CancelPendingFlush itself.

```suggestion
<xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> supports passing a <xref:System.Threading.CancellationToken>. Passing a `CancellationToken` results in an `OperationCanceledException` if the token is canceled while there's a flush pending. `PipeWriter.FlushAsync` supports a way to cancel the current flush operation via <xref:System.IO.Pipelines.PipeWriter.CancelPendingFlush%2A?displayProperty=nameWithType> without raising an exception. Calling `PipeWriter.CancelPendingFlush` causes the current or next call to `PipeWriter.FlushAsync` or `PipeWriter.WriteAsync` to return a <xref:System.IO.Pipelines.FlushResult> with `IsCanceled` set to `true`. This can be useful for halting the yielding flush in a non-destructive and non-exceptional way.
```

<a name="pwcp"></a>

### PipeWriter common problems

* <xref:System.IO.Pipelines.PipeWriter.GetSpan%2A> and <xref:System.IO.Pipelines.PipeWriter.GetMemory%2A> return a buffer with at least the requested amount of memory. **Don't** assume exact buffer sizes.
* There's no guarantee that successive calls will return the same buffer or the same-sized buffer.

> * A new buffer must be requested after calling <xref:System.IO.Pipelines.PipeWriter.Advance%2A> to continue writing more data. The previously acquired buffer can't be written to.

Reviewer 2: This seems understated to me.

The *very start* of the PipeReader section calls out how "It's illegal to use `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo`" and how "Using `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` is undefined". And then it's mentioned again in the PipeReader common problems section. Whereas this, which is basically the same issue with the same risks of memory corruption (if the data was also flushed), is only mentioned in bullet point 3 in the PipeWriter gotchas.

* Calling `GetMemory` or `GetSpan` while there's an incomplete call to `FlushAsync` isn't safe.
* Calling `Complete` or `CompleteAsync` while there's unflushed data can result in memory corruption.

## IDuplexPipe

> The <xref:System.IO.Pipelines.IDuplexPipe> is a contract for types that support both reading and writing. For example, a network connection would be represented by an `IDuplexPipe`.

Reviewer 2: I might add that unlike Pipe which also contains a PipeReader and a PipeWriter, IDuplexPipe represents a single side of a full duplex connection. That means what is written to the PipeWriter will not be read from the PipeReader.

* Passing the wrong values to `consumed` or `examined` may result in infinite buffering (eventual OOM). For example, `PipeReader.AdvanceTo(buffer.Start, buffer.End)` unconditionally when processing a single message at a time from the buffer.
* Using the `ReadOnlySequence<byte>` after calling `PipeReader.AdvanceTo` may result in memory corruption (use after free).
* Failing to call `PipeReader.Complete/CompleteAsync` may result in a memory leak.
* Checking <xref:System.IO.Pipelines.ReadResult.IsCompleted?displayProperty=nameWithType> and exiting the reading logic before processing the buffer results in data loss. The loop exit condition should be based on `ReadResult.Buffer.IsEmpty` and `ReadResult.IsCompleted`. Doing this in the wrong order could result in an infinite loop.

#### Problematic code

❌ **Data loss**

The `ReadResult` can return the final segment of data when `IsCompleted` is set to `true`. Not reading that data before exiting the read loop will result in data loss.

[!INCLUDE [pipelines-do-not-use-1](../../../includes/pipelines-do-not-use-1.md)]

[!code-csharp[DoNotUse#1](~/samples/snippets/csharp/pipelines/DoNotUse.cs?name=snippet)]

> [!INCLUDE [pipelines-do-not-use-2](../../../includes/pipelines-do-not-use-2.md)]

Reviewer 4: do we really need a before and after note?

Author: re: bad code includes before/after
> do we really need a before and after note?

@mairaw  for bad code examples, my vote is yes. People google on API's and copy/paste the first hit sample they find.

BEtter would be to remove the copy button AND put a big red X through the code and disable copy.

Reviewer 4: Ok, I posted on the support channel about this. Hopefully, they'll create a customer suggestion based on that.
